### DISCUSSION TITLE
Eq TypeBound should be removed; Affine could be added.

### DISCUSSION THREAD

**Discussion by ss2165** [Quantinuum/hugr#1298] [Specification]
Currently there is no conceptual coherence in what a TypeBound is. Eq is the smallest bound, but it is defined as "can have equality operation applied to it". This appears to be a special case "interface", much like "hashable", "collection" or "qubit-like" might be. See #842. 

Proposal is instead to restrict TypeBound to be something that describes port connectivity specifically:

**Any**: cannot be implicitly discarded or copied (linear), i.e. exactly one connection from outgoing value port to incoming
[OPTIONAL] **Affine**: can be implicitly discarded, 0 or 1 connections.
**Copyable**: Any number of connections valid. 

These three still form valid subsets of each other, but this is not a strict requirement. 
I mostly just want to remove Eq, the additions may not be worthwhile.

Eq etc. should be considered as part of the interface system, whatever that turns out to be.

Extreme alternative (too extreme imho): 
No type bounds, explicit copy operations, linear/affine are also now part of the interface/constraint system.

**Comment by mark-koch** [comment]
What about ports that must be used at least once (i.e. values can be implicitly copied but not implicitly discarded)? I think you would call these *relevant types* from [relevant logic](https://en.wikipedia.org/wiki/Relevance_logic).

I don't see an immediate application of this though and it would break the nice subset relation we have atm

**Comment by cqc-alec** [comment]
Broadly agree (no Eq). Not entirely opposed to your extreme alternative either.

Couple of points:

- I think the the linear ordering ("these three still form valid subsets of each other") is unimportant. For example if we added Dupable the 4 classes wouldn't be linearly ordered (Affine and Dupable aren't comparable), but this is fine.
- Unless I misunderstand Dupable, this is effectively what Mark calls "relevant"? If we can duplicate, we can copy any positive number of times?

**Comment by acl-cqc** [comment]
Dupable implies both a `dup` and a `free` IIUC, so is just an explicit version of Copyable. Agreed that's about interfaces.

Also not opposed to no-typebounds-and-explicit-copies-everywhere, that'd be a *big* change tho....I think that might be a different discussion....

Mark makes a good point about "Relevant". I don't think sub/superset-ness is vital either, so we could have the full diamond (Any - (Relevant | Affine) - Copyable). OTOH that might also be a good argument for dropping Affine...however I think it's more likely that we'll have affine than relevant. (WTF can be copied but not removed? If you copy it then you now have *two* instances to work out what to do with....)

**Reply by cqc-alec** [reply]
> Dupable implies both a dup and a free IIUC, so is just an explicit version of Copyable

Right, so there isn't actually a distinction between these except that Dupable has specially declared operations for 0-ary and 2-ary copy. Is this really enough reason for Dupable to be its own thing?

> WTF can be copied but not removed?

Hmm. I dunno ... non-erasable memory?

**Reply by acl-cqc** [reply]
> WTF can be copied but not removed?

I now can't get out of my head...the magic brooms in the (original) Sorceror's Apprentice...

**Comment by ss2165** [comment]
Including Dupable was a mistake - it doesn't fit the description of "only pertains to connectivity" because it comes under "Interfaces/TypeClass". Removing for clarity.

**Reply by acl-cqc** [reply]
Yes, fair. Dupable is *related* (but not the same issue) in two ways:
* If we went with "all ports linear, all copies explicit", then Dupable is what the current Copyable looks like
* Dupable relates to type classes with defined operations as it puts names on "special" operations (that are normally represented by edge arity/multiports rather than by nodes - that is: a node is not an op, a node has an op / is an instruction to run an op, and we can also request "free" or "dup" by having 0 or many out-edges)

**Comment by cqc-alec** [comment]
How about using the term "Linear" instead of "Any"? It seems clearer to me; "Any" suggests no constraints, which is really what "Copyable" is.

I don't know how standard the terminology is in [this article](https://en.wikipedia.org/wiki/Substructural_type_system), but it uses "Linear" rather than "Any", and "Normal" rather than "Copyable".

**Reply by acl-cqc** [reply]
Linear instead of Any might be OK? It's fine to treat Copyable types linearly, just you don't have to. (IIRC we've mentioned `Linear == Any - Copyable`, in the spec)

Dislike Copyable -> Any, though. It's a description of what types there are in the set, not what operations/edge-arities it supports. Perhaps the ambiguity between those two senses of "Any" is enough reason not to use Any ?

**Comment by ss2165** [comment]
There are a number of outcomes here that I will order in what I believe to be highest to least consensus. Mercifully each is independent and can be tackled independently.

- [x] #1351 
- [ ] #1356 
- [ ] Add `TypeBound::Affine`
- [ ] Add `TypeBound::Dupable`
- [ ] Add `TypeBound::Relevant`

**Comment by acl-cqc** [comment]
I'm unconvinced about adding Affine. (Ok, middle level of consensus.) I think fewer TypeBounds is better...and really all these proposals (and even the existing TypeBounds) are partial solutions / workarounds to what we really need, which is a system for [type-classes / interfaces](#842). Of course, the fallback way of doing interfaces, is to pass function pointers everywhere (e.g. passing a `copy` or `discard` function on top of the least-capable type i.e. Any/Linear), and even this is not soooo bad for an IR, IMHO.
